Storage/persister/sherlock.h

```diff
@@ -120,66 +132,100 @@ class SherlockStreamPersisterImpl {
         transaction.mutations.emplace_back(BypassVariantTypeCheck(), std::move(entry));
       }
       std::swap(transaction.meta, journal.transaction_meta);
-      stream_used_.Publish(std::move(transaction));
+      Value(publisher_used_)
+          ->template Publish<current::locks::MutexLockStatus::AlreadyLocked>(std::move(transaction), timestamp);
+      PushTimestampToFromLockedSection(timestamp);
     }
     journal.Clear();
   }
 
   void ExposeRawLogViaHTTP(uint16_t port, const std::string& route) {
-    handlers_scope_ +=
-        HTTP(port).Register(route, URLPathArgs::CountMask::None | URLPathArgs::CountMask::One, stream_used_);
+    handlers_scope_ += HTTP(port).Register(route,
+                                           URLPathArgs::CountMask::None | URLPathArgs::CountMask::One,
+                                           [this](Request r) { (*Borrowed<stream_t>(stream_))(std::move(r)); });
   }
 
-  sherlock_t& InternalExposeStream() { return stream_used_; }
+  Borrowed<stream_t> BorrowStream() const { return stream_; }
+  const WeakBorrowed<stream_t>& Stream() const { return stream_; }
+  WeakBorrowed<stream_t>& Stream() { return stream_; }
 
-  template <current::locks::MutexLockStatus MLS>
-  void AcquireDataAuthority() {
-    current::locks::SmartMutexLockGuard<MLS> lock(storage_mutex_ref_);
-    if (stream_used_.DataAuthority() == current::sherlock::StreamDataAuthority::Own) {
-      TerminateStreamSubscription();
-      SyncReplayStream<current::locks::MutexLockStatus::AlreadyLocked>(subscriber_->next_replay_index_);
-      subscriber_ = nullptr;
+  template <current::locks::MutexLockStatus MLS = current::locks::MutexLockStatus::NeedToLock>
+  Borrowed<typename stream_t::publisher_t> PublisherUsed() {
+    locks::SmartMutexLockGuard<MLS> lock(master_follower_change_mutex_);
+    CURRENT_ASSERT(Exists(publisher_used_));
+    return Value(publisher_used_);
+  }
+
+  // Note: `BecomeMasterStorage` can not be called from a locked section, as terminating the subscriber
+  // is by itself an operation that locks the publishing mutex of the stream -- in the destructor.
+  void BecomeMasterStorage() {
+    std::lock_guard<std::mutex> master_follower_change_lock(master_follower_change_mutex_);
+    if (Exists(publisher_used_)) {
+      CURRENT_THROW(StorageIsAlreadyMasterException());
     } else {
-      CURRENT_THROW(UnderlyingStreamHasExternalDataAuthorityException());
+      TerminateStreamSubscriptionFromLockedSection();
+      std::lock_guard<std::mutex> lock(stream_publishing_mutex_ref_);
+      publisher_used_ = nullptr;
+      publisher_used_ = stream_->template BecomeFollowingStream<current::locks::MutexLockStatus::AlreadyLocked>();
```

Comment:
Very good question. TL;DR: It's a naming issue.

`BecomeFollowingStream` on the stream is, in essence, just releasing this stream's publisher to an external authority. I've debated a lot re. this name, and concluded, perhaps wrongly, "become follower stream" is the shortest way to convey the message.

In a way, storage becoming a master storage _impllies_ its underlying stream becomes a following stream -- as this underlying stream's publisher is being released to be exclusively owned by the storage itself.

After some thinking now, it still looks clean and simple (although possibly controversial at the first glance) to me. Perhaps we could add a comment here. Suggestions warmly welcome.